### PR TITLE
[ENKINS-60304] Error handling corrected, retry added on all API calls

### PR DESCRIPTION
**Pull request by lholota**
This fixes a reported bug of incorrect error handling: https://issues.jenkins-ci.org/browse/JENKINS-60304

What has been changed:
- All classes which are used to call GitLab APIs have been wrapped in retry logic
- All places using the GitLabApi adjusted to use the wrappers with retry
- New field added on the UI so that users can change the number of retries (default value is 3 given that this feature is backward compatible)
- Exception handling in GitLabSCMSource changed to throw exceptions out of the methods to let Jenkins know that the scan has failed (if the exception is not thrown, Jenkins treats the empty list as successful result and thinks all branches and MRs have been deleted from the repository)

### CODE REVIEW

**Comment by casz** [comment]
I don't think this is the right place to introduce retry.

**Comment by baymac** [comment]
@lholota Would you be interested in working on this in gitlab4j repository. Once it is merged to that repo we will be able to upgrade it in our plugin.

**Comment by gmessner** [comment]
@casz 
First I would have to understand the need to retry. Please enlighten me.

I see from the retry logic in ```RetryHelper``` that reties are only being done on HTTP statuses of 429 (too many requests) and 500 (internal server error).  Are you running into rate-limiting in testing or actual use?  I had to abandon testing against gitlab.com because of rate-limiting, GitLab4J-API now performs integration testing by using GitLab-CE running in a Docker container and the rate-limiting is no longer a problem.

With the 500 error, retrying could possibly be adding to the problem the server is having.

**Comment by casz** [comment]
I would have to defer to @lholota however I assume users want retry logic much like we have retry logic in GitHub and Bitbucket branch source plugins:
In bitbucket we use the stanrdard http retry logic:
https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/a95edd77f1736fa85af744c55f56681bfb0305da/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L192

GitHub they constantly check against rate limit:
https://github.com/jenkinsci/github-branch-source-plugin/blob/23c8a226eef074da7b87bc4b629f6a9f75bf4766/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L941

Here is GitLab.com current rate limit doc:
https://docs.gitlab.com/ee/user/gitlab_com/index.html#gitlabcom-specific-rate-limits

**Comment by casz** [comment]
To go further into detail with the actual use is that some companies might have +100 to +1000 repos with even more merge requests and branches flying around their repos creating a bunch of Jenkins jobs that all consume against the rate limit.

**Comment by gmessner** [comment]
@casz 
Retrying when a rate limit has been hit seems counter-intuitive, do you have logs or other evidence that retrying actually helps when the rate limit is hit?  From my experience, it actually makes the problem worse.

**Comment by casz** [comment]
@lholota and @guillaumelecerf I would prefer if someone could answer @gmessner and my question whether this belongs here.

File: src/main/resources/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer/config.groovy [hunk 1]
```diff
@@ -16,6 +16,10 @@ f.entry(title: _("Server URL"), field: "serverUrl", "description": "The url to t
     f.textbox(default: GitLabServer.GITLAB_SERVER_URL)
 }
 
+f.entry(title: _("Retry Count"), field: "retryCount", "description": "How many times the API calls to GitLab server should be retried") {
+    f.textbox(default: GitLabServer.RETRY_COUNT)
```
Comment (casz) [inline src/main/resources/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer/config.groovy]:
```suggestion
    f.number(class: 'positive-number', min: 1, step: 1 default: GitLabServer.RETRY_COUNT)
```

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 2]
```diff
@@ -258,6 +275,14 @@ public static FormValidation doCheckServerUrl(@QueryParameter String serverUrl)
...
         }
 
+        public static FormValidation doCheckRetryCount(@QueryParameter Integer retryCount){
+            if(retryCount < 0) {
+                return FormValidation.error(Messages.GitLabServer_invalidRetryCount());
+            }
+
+            return FormValidation.ok();
+        }
+
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]: You can remove this as it is handled by `f.number`

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 3]
```diff
@@ -148,6 +154,17 @@ public String getServerUrl() {
...
     }
 
+    @DataBoundSetter
+    public void setRetryCount(int retryCount) {
+        if(retryCount >= 0) {
+            this.retryCount = retryCount;
+        }
+    }
+
+    public Integer getRetryCount() {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]:
```suggestion
    public int getRetryCount() {
```

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 4]
```diff
@@ -148,6 +154,17 @@ public String getServerUrl() {
         return serverUrl;
     }
 
+    @DataBoundSetter
+    public void setRetryCount(int retryCount) {
+        if(retryCount >= 0) {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]:
```suggestion
```

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 5]
```diff
@@ -148,6 +154,17 @@ public String getServerUrl() {
...
+            this.retryCount = retryCount;
+        }
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]:
```suggestion
```

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 6]
```diff
@@ -79,6 +80,11 @@
     @NonNull
     private final String name;
 
+    /**
+     * How many times each API call to GitLab is retried.
+     */
+    private Integer retryCount;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]:
```suggestion
    private int retryCount;
```

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 7]
```diff
@@ -56,6 +56,7 @@
      * Used as default token value if no any credentials found by given credentialsId.
      */
     public final static String EMPTY_TOKEN = "";
+    public final static Integer RETRY_COUNT = 3;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]:
```suggestion
    public final static int RETRY_COUNT = 3;
```

File: src/main/resources/io/jenkins/plugins/gitlabserverconfig/servers/Messages.properties [hunk 8]
```diff
@@ -4,6 +4,7 @@ GitLabServer.hookManagementAs=Hook management will be performed as {0}
 GitLabServer.validateInterrupted=Interrupted while trying to validate credentials
 GitLabServer.versionInterrupted=Interrupted while trying to determine server version
 GitLabServer.invalidUrl=Invalid GitLab Server URL: {0}
+GitLabServer.invalidRetryCount=Retry count must be greater or equal to 0
```
Comment (casz) [inline src/main/resources/io/jenkins/plugins/gitlabserverconfig/servers/Messages.properties]:
```suggestion
```

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/TagsApiRetryWrapper.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.List;
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.TagsApi;
+import org.gitlab4j.api.models.Tag;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class TagsApiRetryWrapper {
+    private final TagsApi _tagsApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/TagsApiRetryWrapper.java]: This is not C# code, please follow Java code standards.
```suggestion
    private final TagsApi tagsApi;
```

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/GitLabApiRetryWrapper.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import org.gitlab4j.api.GitLabApi;
+import org.gitlab4j.api.GitLabApiException;
+
+public class GitLabApiRetryWrapper {
+    public static final Logger LOGGER = Logger.getLogger(GitLabApiRetryWrapper.class.getName());
+
+    private final GitLabApi _wrappedApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/GitLabApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/GroupApiRetryWrapper.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.GroupApi;
+import org.gitlab4j.api.models.Group;
+import org.gitlab4j.api.models.GroupProjectsFilter;
+import org.gitlab4j.api.models.Project;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class GroupApiRetryWrapper {
+    private final GroupApi _groupApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/GroupApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/MergeRequestApiRetryWrapper.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.List;
+import org.gitlab4j.api.Constants;
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.MergeRequestApi;
+import org.gitlab4j.api.models.MergeRequest;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class MergeRequestApiRetryWrapper {
+    private final MergeRequestApi _mergeRequestApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/MergeRequestApiRetryWrapper.java]: Ditto not following Java code standards

File: .idea/codeStyles/Project.xml [hunk 13]
```diff
@@ -1,133 +0,0 @@
-<component name="ProjectCodeStyleConfiguration">
```
Comment (casz) [inline .idea/codeStyles/Project.xml]: Please do not delete the code standard we added for Intellij

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/CommitsApiRetryWrapper.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import org.gitlab4j.api.Constants;
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.models.Comment;
+import org.gitlab4j.api.models.Commit;
+import org.gitlab4j.api.models.CommitStatus;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class CommitsApiRetryWrapper {
+    private final CommitsApi _commitsApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/CommitsApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/NotesApiRetryWrapper.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package io.jenkins.plugins.gitlabbranchsource.retry;
+
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.NotesApi;
+import org.gitlab4j.api.models.Note;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class NotesApiRetryWrapper {
+    private final NotesApi _notesApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/NotesApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/ProjectApiRetryWrapper.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.ProjectApi;
+import org.gitlab4j.api.models.Member;
+import org.gitlab4j.api.models.Project;
+import org.gitlab4j.api.models.ProjectFilter;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class ProjectApiRetryWrapper {
+    private final ProjectApi _projectApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/ProjectApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/RepositoryApiRetryWrapper.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+import org.gitlab4j.api.RepositoryApi;
+import org.gitlab4j.api.models.Branch;
+import org.gitlab4j.api.models.TreeItem;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class RepositoryApiRetryWrapper {
+    public static final Logger LOGGER = Logger.getLogger(RepositoryApiRetryWrapper.class.getName());
+
+    private final RepositoryApi _repositoryApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/RepositoryApiRetryWrapper.java]: Ditto not following Java code standards

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/RepositoryFileApiRetryWrapper.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.io.InputStream;
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.RepositoryFileApi;
+import org.gitlab4j.api.models.RepositoryFile;
+
+import static io.jenkins.plugins.gitlabbranchsource.retry.RetryHelper.withRetry;
+
+public class RepositoryFileApiRetryWrapper {
+    private final RepositoryFileApi _repositoryFileApi;
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/retry/RepositoryFileApiRetryWrapper.java]: Ditto not following Java code standards

**Comment by gmessner** [comment]
@casz 

> @gmessner
> I still believe that retry logic should be added to gitlab4j.
> 
> Retry logic is standard in REST apis.
> 
> See
> 
> https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/a95edd77f1736fa85af744c55f56681bfb0305da/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L192
> 
> and
> 
> https://github.com/jenkinsci/github-branch-source-plugin/blob/3d1702dbd290ae9cf872ee65b47b16b105ae9032/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L1674

As I have stated previously retrying when hitting a rate limit can exacerbate the problem.  And retrying when a 500 error is returned also has the potential to make the problem worse on the server.  This is why I've asked for some evidence (logs) that retrying when gitlab.com returns a 429 or 500 error actually ever produces a non-error response.  

The best ways I have found to mitigate hitting rate limits at gitlab.com was to examine what API requests were being made and finding ways to reduce the number of requests but still get the required information and to cache information at the application level when possible.

That being said, when I have some time (I am quite busy right now) I will look into adding retry logic, but will only add it if testing demonstrates that retrying has some positive results.

**Comment by casz** [comment]
@gmessner with the right gitlab.com retry logic you can wait for the reset point:

https://docs.gitlab.com/ee/user/gitlab_com/index.html#haproxy-api-throttle

**Comment by gmessner** [comment]
@casz 
I completely understand how to calculate the retry wait time,  I've actually done this in the past at the app level,  but from experience, it rarely did any good, which is why I was asking for some evidence that retrying actually has been successful. 

BTW, it seems that the request has changed to only retrying when a 429 error has occurred, this would be good because retrying when a 500 or other server-side errors occur are typically a bad practice.

**Comment by casz** [comment]
I do not use gitlab.com so I cannot provide any details, hence why I asked others to provide them.

**Comment by baymac** [comment]
@lholota Thank you for your contribution. I am closing this for now as we might not be going this path. Please let me know if you still want to merge it into the plugin.

**Comment by baymac** [comment]
@lholota Are you interested in picking this up? After my experience some other projects, I feel the retry logic should be implemented by the client than the API itself. I would interested to work on this but would be great if you pick it up.

**Comment by jpraet** [comment]
I am also impacted by https://issues.jenkins-ci.org/browse/JENKINS-60304. Our artifact versioning scheme makes use of the jenkins build number, and when there is a connection error when scanning for branches, the jobs are recreated causing the build number sequence to reset. 

All remarks on this PR are related to the retry logic that was added. Could the actual error handling bugfix maybe get patched seperately?

**Comment by redeamer** [comment]
As we hit this issue too on every? update of gitlab, and restoring of the build-numbers for dev/release/master branches is bothersome for all gitlab projects, I created another PR which only addresses the Exception 'forwarding': #88

**Pull request closed without merging**